### ISSUE TITLE
coder2 gpg agent cleanup

### ISSUE THREAD

**Issue by jkozlowski** [coder/coder#12156]
I have been using coder2 almost exclusively over the last week, so my workspaces are open all day long. Usually when I come back after a period of inactivity, the terminal I have open to forward GPG things, so I can use my yubikey, is unresponsive. So I do what I'd usually do, close that terminal and rerun the command. When I do that, it hangs for a while and then prints:

```
➜ coder ssh -G myworkspace
Encountered an error running "coder ssh"
upload GPG public keys and ownertrust to workspace: check if agent socket is running (check if "+ gpgconf --list-dir agent-socket\n/home/coder/.gnupg/S.gpg-agent\n+ agent_socket=/home/coder/.gnupg/S.gpg-agent\n+ echo /home/coder/.gnupg/S.gpg-agent\n+ [ -S /home/coder/.gnupg/S.gpg-agent ]\n+ echo agent socket exists, attempting to kill it\nagent socket exists, attempting to kill it\n+ gpgconf --kill gpg-agent\ngpgconf: error running '/usr/bin/gpg-connect-agent': terminated\ngpgconf: error running '/usr/bin/gpg-connect-agent KILLAGENT': General error" exists): `sh -c '
set -eux
agent_socket=$(gpgconf --list-dir agent-socket)
echo "$agent_socket"
if [ -S "$agent_socket" ]; then
  echo "agent socket exists, attempting to kill it" >&2
  gpgconf --kill gpg-agent
  rm -f "$agent_socket"
  sleep 1
fi

test ! -S "$agent_socket"
'` failed: stderr:

stdout: + gpgconf --list-dir agent-socket
/home/coder/.gnupg/S.gpg-agent
+ agent_socket=/home/coder/.gnupg/S.gpg-agent
+ echo /home/coder/.gnupg/S.gpg-agent
+ [ -S /home/coder/.gnupg/S.gpg-agent ]
+ echo agent socket exists, attempting to kill it
agent socket exists, attempting to kill it
+ gpgconf --kill gpg-agent
gpgconf: error running '/usr/bin/gpg-connect-agent': terminated
gpgconf: error running '/usr/bin/gpg-connect-agent KILLAGENT': General error:

Process exited with status 1: Process exited with status 1
```

If I manually restart the local and remote agent, I can then usually successfully run the command and get a session established (even if sometimes it hangs for a while). How can I debug what's happening? I suspect it's some weird ssh thing, but it's actively preventing me from using your product successfully

**Comment by deansheather** [comment]
I think this is due to the agent socket being left behind when the connection is improperly closed. We have logic (added last month in 385d58caf6347e48657d4fbe0211c6f84cde7199) to allow for re-listening on those sockets within the same agent "run", but they might still get left behind until we attempt to re-listen on them again.

We most likely need to change this script to ignore errors from `gpgconf --kill gpg-agent` and then `rm` the socket if it's still there afterwards. Will need some testing though.

For now, if you SSH in without `-G`, then run `rm /home/coder/.gnupg/S.gpg-agent`, then SSH in again with `-G` does it work?

**Comment by jkozlowski** [comment]
Yep, I can confirm that’s how I worked around this. It sounds like you’re already working on this, feel free to close

**Comment by jkozlowski** [comment]
were there any fixes here recently? qualitatively speaking, I haven't observed this problem in a while

**Comment by deansheather** [comment]
That's great to hear. I believe some stuff around sockets was changed in the last few months which may have fixed the issue. I'll close this issue and if you encounter it again feel free to reopen :)
